### PR TITLE
Add the functionality to merge and unmerge MessageLoopTaskQueues

### PR DESCRIPTION
**Pull request by iskakaushik**
This introduces a notion of a "owning" and "subsumed" queue ids.
Owning queue will take care of the tasks submitted to both that and it's
subsumed queue.

- The tasks submitted still maintain the queue affinity
- Same for the task observers

- Also adds MergedQueuesRunner which grabs both the locks owner
  and subsumed queues in RAII fashion.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,5 +21,6 @@ Thumbs.db
 .idea
 pubspec.lock
 .vscode/
+compile_commands.json
```
Comment (chinmaygarde) [inline .gitignore]: Compile commands should only end up in `out/`. This should not be necessary.

File: fml/merged_queues_runner.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include "flutter/fml/message_loop_task_queues.h"
+
+namespace fml {
+
+// RAII class for managing merged locks.
+class MessageLoopTaskQueues::MergedQueuesRunner {
+ public:
+  // TODO (kaushikiska): refactor mutexes out side of MessageLoopTaskQueues
+  // for better DI.
+  MergedQueuesRunner(MessageLoopTaskQueues* task_queues,
```
Comment (chinmaygarde) [inline fml/merged_queues_runner.cc]: Since the queue can't ever be nullptr, use a reference for clarity.

File: fml/message_loop_task_queues.h [hunk 3]
```diff
@@ -104,6 +134,11 @@ class MessageLoopTaskQueues
   std::vector<TaskObservers> task_observers_;
   std::vector<DelayedTaskQueue> delayed_tasks_;
 
+  static const size_t _kUnmerged = ULONG_MAX;
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.h]: Lets typedef size_t

File: fml/message_loop_task_queues.cc [hunk 4]
```diff
@@ -140,9 +164,136 @@ void MessageLoopTaskQueues::Swap(TaskQueueId primary, TaskQueueId secondary)
...
+bool MessageLoopTaskQueues::Merge(TaskQueueId owner, TaskQueueId subsumed) {
+  // task_observers locks
+  std::mutex& o1 = GetMutex(owner, MutexType::kObservers);
+  std::mutex& o2 = GetMutex(subsumed, MutexType::kObservers);
+
+  // delayed_tasks locks
+  std::mutex& t1 = GetMutex(owner, MutexType::kTasks);
+  std::mutex& t2 = GetMutex(subsumed, MutexType::kTasks);
+
+  std::scoped_lock(o1, o2, t1, t2);
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.cc]: This does not actually perform the lock. Please use a variable here.

File: fml/message_loop_task_queues.cc [hunk 5]
```diff
@@ -140,9 +164,136 @@ void MessageLoopTaskQueues::Swap(TaskQueueId primary, TaskQueueId secondary)
...
+}
+
+bool MessageLoopTaskQueues::Unmerge(TaskQueueId owner) {
+  // task_observers locks
+  std::mutex& o = GetMutex(owner, MutexType::kObservers);
+
+  // delayed_tasks locks
+  std::mutex& t = GetMutex(owner, MutexType::kTasks);
+
+  std::scoped_lock(o, t);
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.cc]: This does not actually perform the lock. Please use a variable here.

File: fml/message_loop_task_queues.cc [hunk 6]
```diff
@@ -140,9 +164,136 @@ void MessageLoopTaskQueues::Swap(TaskQueueId primary, TaskQueueId secondary)
...
+
+  return true;
+}
+
+bool MessageLoopTaskQueues::Unmerge(TaskQueueId owner) {
+  // task_observers locks
+  std::mutex& o = GetMutex(owner, MutexType::kObservers);
+
+  // delayed_tasks locks
+  std::mutex& t = GetMutex(owner, MutexType::kTasks);
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.cc]: Not acquiring the subsumed lock will cause unsafe access to the tasks/observers heap in `HasPendingTasksUnlocked(subsumed)`. Just acquire all four locks.

File: fml/message_loop_task_queues.cc [hunk 7]
```diff
@@ -52,39 +58,47 @@ void MessageLoopTaskQueues::RegisterTask(TaskQueueId queue_id,
...
+  while (HasPendingTasksUnlocked(queue_id)) {
+    TaskQueueId top_queue;
+    const auto& top = PeekNextTaskUnlocked(queue_id, top_queue);
     if (top.GetTargetTime() > now) {
       break;
     }
     invocations.emplace_back(std::move(top.GetTask()));
-    tasks.pop();
+    delayed_tasks_[top_queue].pop();
     if (type == FlushType::kSingle) {
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.cc]: We no longer make use of this in the concurrent tasks queue. So lets just get rid of flush type with the default being multiple.

File: fml/message_loop_task_queues.cc (hunk 4 above)
Comment (chinmaygarde) [reply fml/message_loop_task_queues.cc]: Apparently I also missed the scoped_lock in the other review. Just global search and replace this pattern in the codebase.

File: fml/merged_queues_runner.cc [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    task_queues_->GetMutex(owner_, type_).unlock();
+  }
+
+ private:
+  bool isMerged(TaskQueueId queue_id) {
+    return queue_id != MessageLoopTaskQueues::_kUnmerged;
+  }
+
+  const TaskQueueId owner_;
+  TaskQueueId subsumed_;
```
Comment (chinmaygarde) [inline fml/merged_queues_runner.cc]: Lets make task queue ID a struct with a single member of `size_t`.

**Comment by chinmaygarde** [comment]
The approach looks good. Have highlighted the thread safety issue and some other improvements.

File: fml/message_loop_task_queues.cc (hunk 7 above)
Comment (iskakaushik) [reply fml/message_loop_task_queues.cc]: Created this https://github.com/flutter/flutter/issues/36083, will follow this up.

File: .gitignore (hunk 1 above)
Comment (iskakaushik) [reply .gitignore]: Done!

File: fml/merged_queues_runner.cc (hunk 2 above)
Comment (iskakaushik) [reply fml/merged_queues_runner.cc]: Done

File: fml/merged_queues_runner.cc (hunk 8 above)
Comment (iskakaushik) [reply fml/merged_queues_runner.cc]: Done

File: fml/message_loop_task_queues.cc (hunk 4 above)
Comment (iskakaushik) [reply fml/message_loop_task_queues.cc]: Done

File: fml/message_loop_task_queues.cc (hunk 6 above)
Comment (iskakaushik) [reply fml/message_loop_task_queues.cc]: Done

File: fml/message_loop_task_queues.cc (hunk 5 above)
Comment (iskakaushik) [reply fml/message_loop_task_queues.cc]: Done

File: fml/message_loop_task_queues.h (hunk 3 above)
Comment (iskakaushik) [reply fml/message_loop_task_queues.h]: Moved this to a struct

File: fml/message_loop_task_queues.h [hunk 9]
```diff
@@ -18,7 +18,15 @@
...
 
-typedef size_t TaskQueueId;
+class TaskQueueId {
+ public:
+  explicit TaskQueueId(size_t value) : value_(value) {}
+
+  operator int() const { return value_; }
+
+ private:
+  size_t value_;
```
Comment (chinmaygarde) [inline fml/message_loop_task_queues.h]: Zero it out to kUnused here by default.

**Pull request merged**